errors: port internal/fs errors to internal/errors

gunar opened the pull request:
* Assign codes to errors reported by internal/fs.js

Ref: #11273

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->
Semver-major because this updates specific error messages and converts errors over to use the new internal/errors.js mechanism.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows commit guidelines

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
errors, fs

joyeecheung commented on lib/internal/errors.js:
```diff
@@ -86,3 +86,5 @@ module.exports = exports = {
 // Note: Please try to keep these in alphabetical order
 E('ERR_ASSERTION', (msg) => msg);
 // Add new errors from here...
+E('ERR_UNK_FILE_OPEN_FLAG', 'Unknown file open flag: %s');
```
Hmm, do we have a preference between `util.format` strings and arrow functions + templates? I think the latter should be more performant though(haven't tested) @jasnell

joyeecheung replied:
Although `util.format` strings should be more readable, and since we are throwing errors performances probably is not the primary concern here..

jasnell replied:
the code is set up to support either. this should be fine.

gunar replied:
Even if both work, for consistency purposes we should commit to either one.

thefourtheye commented on lib/internal/fs.js:
```diff
@@ -17,7 +18,7 @@ const O_WRONLY = constants.O_WRONLY | 0;
 function assertEncoding(encoding) {
   if (encoding && !Buffer.isEncoding(encoding)) {
-    throw new Error(`Unknown encoding: ${encoding}`);
+    throw new errors.Error('ERR_UNK_ENCODING', encoding)
```
Semicolon is missing at the end.

gunar replied:
Fixed.

TimothyGu commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/6366/

gunar commented:
1. At what moment should I squash my commits?
2. Is there anything lacking in this PR as it is?

Thank you!

joyeecheung commented:
@gunar I think this can be squashed by whoever lands this. This PR needs a CITGM run though I am not sure if the CITGM CI works at the moment?

fhinkel commented:
@gunar Thanks so much for putting this together. Sorry that it is dragging out for so long due to being a semver-major change. Could you rebase and also squash your commits (I think all the changes should be one commit, right?). Thanks!

fhinkel commented:
I'm closing this because it's been inactive for quite a while. Feel free to reopen or ping a collaborator to get it reopened if needed.

gunar commented:
Rebased, squashed, pushed to https://github.com/gunar/node/tree/internal-errors-fs.
Ready for review.
However, this PR did not get updated with the newest commits.
Guess it'll happen as soon as we reopen the PR.
@fhinkel Would you reopen it, please? Thanks

joyeecheung commented:
@gunar This PR can not be reopened with a rebased branch, you can follow https://github.com/isaacs/github/issues/361#issuecomment-219088644 to reopen it by moving the branch to where it was first, then reopen, then rebase again.

gunar commented:
@joyeecheung Gotcha. Should be good to go now. Please reopen.

mhdawson commented:
reopened.

BridgeAR commented on lib/internal/fs.js:
```diff
@@ -52,7 +53,7 @@ function stringToFlags(flag) {
   }
 
-  throw new Error('Unknown file open flag: ' + flag);
+  throw new errors.Error('ERR_UNK_FILE_OPEN_FLAG', flag);
```
I recommend to use the `ERR_INVALID_OPT_VALUE` error type for both cases here instead of creating new ones.

refack replied:
I second.
IMHO the special case is little bit counterproductive:
```js
try {
  // do some stuff
  var f = fs.openSync('g.txt', 'rgh');
  // do other stuff
catch (e) {
  if (e.code == 'ERR_INVALID_OPT_VALUE') ; // swallow because I'm a special kind of guy
  else throw e;
}
```

gunar replied:
Agreed. Done.

BridgeAR commented on test/parallel/test-fs-assert-encoding-error.js:
```diff
@@ -4,72 +4,72 @@ const assert = require('assert');
 
 const options = 'test';
-const unknownEncodingMessage = /^Error: Unknown encoding: test$/;
+const expectedError = common.expectsError('ERR_INVALID_OPT_VALUE');
```
Please also check for the type as this is not done anymore with the current check. Please change this in all the test files accordingly.

gunar commented:
I think this is what you meant
Let me know otherwise

If all is good I'll squash the commits
Thanks!

BridgeAR commented on test/parallel/test-fs-assert-encoding-error.js:
```diff
@@ -4,72 +4,72 @@ const assert = require('assert');
 
 const options = 'test';
-const unknownEncodingMessage = /^Error: Unknown encoding: test$/;
+const expectedError = common.expectsError({ type: TypeError, code: 'ERR_INVALID_OPT_VALUE' });
```
The line length exceeds 80 chars in all tests

joyeecheung commented on lib/internal/fs.js:
```diff
@@ -52,7 +53,7 @@ function stringToFlags(flag) {
   }
 
-  throw new Error('Unknown file open flag: ' + flag);
+  throw new errors.Error('ERR_INVALID_OPT_VALUE', 'flags', flag);
```
nit: drop the `s` in `flags` because it's `flag` that is in the options

joyeecheung commented on test/parallel/test-fs-assert-encoding-error.js:
```diff
@@ -4,72 +4,73 @@ const assert = require('assert');
 const options = 'test';
-const unknownEncodingMessage = /^Error: Unknown encoding: test$/;
+const expectedError =
+  common.expectsError({code: 'ERR_INVALID_OPT_VALUE', type: TypeError});
```
Either this should be `Error` or the error thrown in `fs` should be `TypeError`..

refack replied:
+1 For `TypeError` in both.

joyeecheung commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/8796/

refack commented:
Quick sanity: https://ci.nodejs.org/job/node-test-commit-linuxone/6821/

refack commented on lib/internal/fs.js:
```diff
@@ -52,7 +53,7 @@ function stringToFlags(flag) {
   }
 
-  throw new Error('Unknown file open flag: ' + flag);
+  throw new errors.TypeError('ERR_INVALID_OPT_VALUE', 'flag', flag);
```
@gunar Sorry to send you back and forth, I just looked at the docs https://nodejs.org/api/fs.html#fs_fs_open_path_flags_mode_callback
IMHO the formal parameter should be `flags` so for consistency do `/s/flag/flags/` for this whole function.
@joyeecheung do you agree?

refack commented:
CI gave us a missed test case:
```js
assert.js:569
    throw actual;
    ^

TypeError [ERR_INVALID_OPT_VALUE]: The value "foo" is invalid for option "encoding"
    at Object.assertEncoding (internal/fs.js:22:11)
    at assert.throws (/data/iojs/build/workspace/node-test-commit-linuxone/nodes/rhel72-s390x/test/parallel/test-internal-fs.js:10:24)
    at _tryBlock (assert.js:526:5)
    at _throws (assert.js:547:12)
    at Function.throws (assert.js:577:3)
    at Object.<anonymous> (/data/iojs/build/workspace/node-test-commit-linuxone/nodes/rhel72-s390x/test/parallel/test-internal-fs.js:10:8)
    at Module._compile (module.js:569:30)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:503:32)
    at tryModuleLoad (module.js:466:12)
```

refack commented:
Sanity: https://ci.nodejs.org/job/node-test-commit-linuxone/6822/ ✔️ 
Linter: https://ci.nodejs.org/job/node-test-linter/10028/ 🔴

refack commented on test/parallel/test-internal-fs.js:
```diff
@@ -8,4 +8,4 @@ const fs = require('internal/fs');
 assert.doesNotThrow(() => fs.assertEncoding('utf8'));
 assert.throws(() =>
-    common.expectsError({code: 'ERR_INVALID_OPT_VALUE', type: TypeError}));
+  common.expectsError({code: 'ERR_INVALID_OPT_VALUE', type: TypeError}));
```
Two comments:
1. you ate the test block `() => fs.assertEncoding('foo')`
2. common is undefined (required but not assigned)

This passed since `() => common.expectsError({code: 'ERR_INVALID_OPT_VALUE', type: TypeError}))` throw `Error Undefined`

gunar replied:
hah, good catch!
I can't find out what's wrong with linting, though

refack replied:
It was that.

refack replied:
Now it's `common is undefined`. Change line 4.

gunar replied:
d'oh

mhdawson commented:
new CI run: https://ci.nodejs.org/job/node-test-pull-request/8811/

refack commented:
@gunar there's a lint error.

jasnell commented on lib/internal/fs.js:
```diff
@@ -18,16 +19,16 @@ const {
 function assertEncoding(encoding) {
   if (encoding && !Buffer.isEncoding(encoding)) {
-    throw new Error(`Unknown encoding: ${encoding}`);
+    throw new errors.TypeError('ERR_INVALID_OPT_VALUE', 'encoding', encoding);
```
Definitely prefer a more specific error for `Unknown encoding`. We throw these in several places throughout the source, often enough that it makes sense to identify them separately.

refack replied:
IMHO if we specialize it should be `ERR_INVALID_OPT_VALUE_ENCODING`

gunar replied:
Makes sense. I don't think I am the one who should make the call. Let me know and I'll implement.

refack replied:
Go with `ERR_INVALID_OPT_VALUE_ENCODING`, I believe @jasnell will agree

gunar replied:
Friendly reminder: We went from `ERR_UNKNOWN_ENCODING` to `ERR_INVALID_OPT_VALUE` to `ERR_INVALID_OPT_VALUE_ENCODING`.

refack replied:
> Friendly reminder: We went from `ERR_UNKNOWN_ENCODING` to `ERR_INVALID_OPT_VALUE` to `ERR_INVALID_OPT_VALUE_ENCODING`.

I'm sorry to have made you run around... The new `Error` concept still has some undecided cases, you are just one of the first to tackle them...

gunar commented:
@refack I can't seem to find the linting error
Shouldn't it be somewhere in the CI page?

refack commented:
> @refack I can't seem to find the linting error
> Shouldn't it be somewhere in the CI page?

You should run `make lint` or `vcbuild nobuild noprojgen lint`

As for the CI, not yet, it's a todo... https://github.com/nodejs/build/issues/720 but if you hit the machine just after it finishes the `tap` file is there:
https://ci.nodejs.org/job/node-test-linter/10114/
```
not ok 33 - /usr/home/iojs/build/workspace/node-test-linter/test/parallel/test-internal-fs.js
  ---
  message: Expected indentation of 14 spaces but found 2.
  severity: error
  data:
    line: 11
    column: 3
    ruleId: indent-legacy
  ...
```

gunar commented:
Check 'em

refack commented:
https://ci.nodejs.org/job/node-test-pull-request/8867/

IMHO the PR is ready so I'm going to land it later today.
Since it's `semver-major` we have a ample time to decide if we want to revamp the error codes during the `v9.x` cooking timeframe...

refack commented:
Missed one:
``` JS
761	parallel/test-internal-fs	
duration_ms	0.59
severity	fail
stack	
assert.js:60
  throw new errors.AssertionError({
  ^

AssertionError [ERR_ASSERTION]: 'ERR_INVALID_OPT_VALUE_ENCODING' === 'ERR_INVALID_OPT_VALUE'
    at Object.<anonymous> (/data/iojs/build/workspace/node-test-commit-linuxone/nodes/rhel72-s390x/test/common/index.js:701:12)
    at expectedException (assert.js:520:19)
    at _throws (assert.js:568:8)
    at Function.throws (assert.js:577:3)
    at Object.<anonymous> (/data/iojs/build/workspace/node-test-commit-linuxone/nodes/rhel72-s390x/test/parallel/test-internal-fs.js:10:8)
    at Module._compile (module.js:569:30)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:503:32)
    at tryModuleLoad (module.js:466:12)
    at Function.Module._load (module.js:458:3)
```

gunar commented:
I did forget one!
Disclaimer: I have Py3 on my machine so getting tests to run locally is tricky.

Let me know when it's good to squash
I rather have separate commits during dev

gunar commented:
Is anything else required of me?
I'd like to get this merged :-)

refack commented:
> Is anything else required of me?
> I'd like to get this merged :-)

I don't see anything else.
Pre-land CI: https://ci.nodejs.org/job/node-test-commit/11225/

refack commented:
@gunar even though the end state has no conflicts with `master` is seems like the intermediate commits do, so could you squash and rebase?

From the CI machine:
```
+ git rebase --committer-date-is-author-date f406a7ebaee09c00b6dec330e17897924096c30d
First, rewinding head to replay your work on top of it...
Applying: errors: port internal/fs errors to internal/errors
Using index info to reconstruct a base tree...
M	doc/api/errors.md
M	lib/internal/errors.js
Falling back to patching base and 3-way merge...
Auto-merging lib/internal/errors.js
CONFLICT (content): Merge conflict in lib/internal/errors.js
Auto-merging doc/api/errors.md
CONFLICT (content): Merge conflict in doc/api/errors.md
Failed to merge in the changes.
Patch failed at 0001 errors: port internal/fs errors to internal/errors
The copy of the patch that failed is found in:
   /var/lib/jenkins/jobs/node-test-commit/workspace@6/.git/rebase-apply/patch

When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".
```

gunar commented:
Sure thing! There you go.

refack commented:
Landed in 81084d47b5

refack commented:
@gunar Congrats on GitHub promoting you from:
![image](https://user-images.githubusercontent.com/96947/28346828-5a520cfa-6c01-11e7-82ca-b0311487ce18.png)
To
![image](https://user-images.githubusercontent.com/96947/28347075-f126d79a-6c02-11e7-8d4c-6f8b5459e87e.png)

refack commented:
Pulled out and re-running CI: https://ci.nodejs.org/job/node-test-commit/11243/

gunar commented:
Siiiick. Thank, man!
